track/gpsd.py

```diff
@@ -0,0 +1,185 @@
+"""GPS location determination via gpsd.
+
+Defines a single class GPS that encapsulates the procedures necessary to obtain the current location
+from the gpsd daemon. This code works on the assumption that gpsd is already configured and running
+on the system.
+"""
+
+# http://manpages.ubuntu.com/manpages/bionic/man5/gpsd_json.5.html
+
+from collections import namedtuple, OrderedDict
+from enum        import Enum, Flag, auto
+from math        import inf, nan, isnan
+from time        import perf_counter
+#from types import SimpleNamespace
+
+import gps
+
+# improved enum-ized version of the 'MODE_' constants from the gps module
+class GPSFixType(Enum):
+    ZERO   = 0
+    NO_FIX = 1
+    FIX_2D = 2
+    FIX_3D = 3
+
+GPSLocation = namedtuple('GPSLocation', ['lat', 'lon', 'alt'])
+GPSErrors   = namedtuple('GPSErrors',   ['lat', 'lon', 'alt', 'track', 'speed', 'climb', 'time'])
+
+# TODO: document the units on all the relevant fields here, and use types for them when appropriate
+#       (ask Brett for the best degree types and unit stuff to use here... astropy has good ones?)
+# - lat: decimal degrees, [-90.0, +90.0]; north positive
+# - lon: decimal degrees, [-180.0, +180.0] or similar (I think); east positive
+# - alt: meters above sea level, [-inf, +inf]
+# - track: decimal degrees, [0.0, 360.0); standard compass heading: clockwise from north
+# - speed: meters per second in XY plane, [0.0, +inf]
+# - climb: meters per second in Z  plane, [-inf, +inf]; upward positive
+# - time: absolute UTC time; up to millisecond precision
+# - errors: same units as the corresponding value; 95% confidence; always nonnegative;
+#           time error is in seconds
+
+# TODO: put this somewhere else where it would belong, or find an existing module that provides this
+def const(val): return property(fget=lambda _: val)
```

Comment:
What does this do? Kinda obtuse

Reply:
It's a geniusly succinct way of making a const value.

Basically it's a property. It has a getter and no setter. And the getter is just a lambda that returns the specified value.

If you attempt to assign to something created with this `const()` thing, you'll get an exception.

Reply:
And the `_` is there because the getter is supposed to take a `self` parameter but I had no need for that.